src/passes/MemoryCopyFillLowering.cpp

```diff
@@ -31,218 +47,199 @@ struct MemoryCopyFillLowering
+                                b.makeLocalGet(end, Type::i32)),
+                   b.makeBinary(BinaryOp::GtUInt32,
+                                b.makeBinary(BinaryOp::AddInt32,
+                                             b.makeLocalGet(src, Type::i32),
+                                             b.makeLocalGet(size, Type::i32)),
+                                b.makeLocalGet(end, Type::i32))),
+      b.makeUnreachable()));
+    // if src < dest
+    body->list.push_back(
+      b.makeIf(b.makeBinary(BinaryOp::LtUInt32,
+                            b.makeLocalGet(src, Type::i32),
+                            b.makeLocalGet(dst, Type::i32)),
+               b.makeBlock({
+                 b.makeLocalSet(start,
+                                b.makeBinary(BinaryOp::SubInt32,
+                                             b.makeLocalGet(size, Type::i32),
+                                             b.makeConst(1))),
+                 b.makeLocalSet(end, b.makeConst(-1U)),
+                 b.makeLocalSet(step, b.makeConst(-1U)),
+               }),
+               b.makeBlock({
+                 b.makeLocalSet(start, b.makeConst(0)),
+                 b.makeLocalSet(end, b.makeLocalGet(size, Type::i32)),
+                 b.makeLocalSet(step, b.makeConst(1)),
+               })));
+    // i = start
+    body->list.push_back(b.makeLocalSet(i, b.makeLocalGet(start, Type::i32)));
+    body->list.push_back(b.makeBlock(
+      "out",
+      b.makeLoop(
+        "copy",
+        b.makeBlock(
+          {// break if i == end
+           b.makeBreak("out",
+                       nullptr,
+                       b.makeBinary(BinaryOp::EqInt32,
+                                    b.makeLocalGet(i, Type::i32),
+                                    b.makeLocalGet(end, Type::i32))),
+           // dst[i] = src[i]
+           b.makeStore(1,
+                       0,
+                       1,
+                       b.makeBinary(BinaryOp::AddInt32,
+                                    b.makeLocalGet(dst, Type::i32),
+                                    b.makeLocalGet(i, Type::i32)),
+                       b.makeLoad(1,
+                                  false,
+                                  0,
+                                  1,
+                                  b.makeBinary(BinaryOp::AddInt32,
+                                               b.makeLocalGet(src, Type::i32),
+                                               b.makeLocalGet(i, Type::i32)),
+                                  Type::i32,
+                                  memory),
+                       Type::i32,
+                       memory),
+           // i += step
+           b.makeLocalSet(i,
+                          b.makeBinary(BinaryOp::AddInt32,
+                                       b.makeLocalGet(i, Type::i32),
+                                       b.makeLocalGet(step, Type::i32))),
+           // loop
+           b.makeBreak("copy", nullptr)}))));
```

Comment:
Reading and writing code written like this is pretty painful.   I know its a little late now but I wonder if we should have just generated calls to function imports and then rely on the embedder/JS providing them?

Reply:
The reason I didn't go with that is because I don't think we can really rely on the linking steps this late in the pipeline. It's similar to the problem of memcpy calls being generated during LTO, but this runs even later.

Reply:
As @kripken suggested above, we could have done our own linking hack by writing wat files, assembling them and then merging them in with wasm-merge, but that moves more complexity into emscripten; it didn't sound much better to me.

Reply:
I know its a bit late in the day here to suggest it but thee is precedent in emscripten for the linker adding extra imports that it know were generated by late binaryen passes.  For example we do it for SAFE_HEAP:

https://github.com/emscripten-core/emscripten/blob/4b3bacecc7efc5eddbe4670046baa43bc002422d/tools/emscripten.py#L781-L783

And for AUDODEBUG:

https://github.com/emscripten-core/emscripten/blob/4b3bacecc7efc5eddbe4670046baa43bc002422d/tools/emscripten.py#L795-L824

Reply:
Maybe not worth changing this now that you've already written all this code?

Reply:
I see, so you're suggesting that the implementation actually be JS, rather than in wasm. I still think that the complexity of such a system (that would require JS library code, a Binaryen pass, and emscripten linking logic to wire everything together) is worse than having a self-contained Binaryen pass that has a bit more complexity in the pass itself. Writing (and even debugging) this pass was almost the easiest thing about bringing bulk-memory on by default; all the hard part was dealing with the complexity in emscripten. 
I dunno, I'm willing to be convinced, but I'm not there yet, curious what you and others think about the complexity there.

Having said all that, this clang-format formatting seems much less readable than what I originally wrote. Not sure if there's anything we can do about that or not.

Reply:
I'm ok with leaving this as is.  You correct that having it self contained is nice.

You can write `// clang-format off` and `// clang-format on` around the block in question.

Reply:
I pushed an improved version, not sure if it's worthwhile or not though.

Reply:
I decided it wasn't enough of an improvement to be worth it; you can check [02201ef](https://github.com/WebAssembly/binaryen/pull/7021/commits/02201ef3d123b35d864b7357a6d928a842e69126) in case you disagree.